### PR TITLE
[TENSORFLOW][NEURON][BUILD][TEST] - Neuron TF2.5

### PR DESCRIPTION
**Pull request by aws-vrnatham**
Add support for tf2.5 neuron

Signed-off-by: Venky Natham <vrnatham@amazon.com>

*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
The neuron team released Tensorflow 2.5 support and add a DLC image with the tensorflow neuron version 2.5

### Tests run
Added new tests for TF2.5

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true` or `neuron_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tensorflow/buildspec-neuron.yml [hunk 1]
```diff
@@ -14,25 +15,31 @@
...
+      deep_learning_container:
+        source: ../../src/deep_learning_container.py
+        target: deep_learning_container.py
 
   images:
     BuildNeuronTFInferencePy3DockerImage:
       <<: *INFERENCE_REPOSITORY
       build: &TENSORFLOW_CPU_INFERENCE_PY3 false
-      image_size_baseline: 5000
+      image_size_baseline: 10000
```
Comment (saimidu) [inline tensorflow/buildspec-neuron.yml:37]: What is the actual size of the image built? Please update this with that size in MB.

File: tensorflow/inference/docker/2.5/py3/sdk1.16.0/Dockerfile.neuron [hunk 2]
```diff
@@ -0,0 +1,152 @@
+FROM ubuntu:18.04
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+# Specify accept-bind-to-port LABEL for inference pipelines to use SAGEMAKER_BIND_TO_PORT
+# https://docs.aws.amazon.com/sagemaker/latest/dg/inference-pipeline-real-time.html
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
```
Comment (saimidu) [inline tensorflow/inference/docker/2.5/py3/sdk1.16.0/Dockerfile.neuron:8]: Is this label valid?

File: tensorflow/inference/docker/2.5/py3/sdk1.16.0/Dockerfile.neuron [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+# using --no-dependencies to avoid installing tensorflow binary
+ && ${PIP} install --no-dependencies --no-cache-dir \
+    tensorflow-serving-api==2.5.1
+
+# pip install statements have been separated out into multiple sequentially executed statements to
+# resolve package dependencies during installation.
+RUN ${PIP} install "tensorflow_serving_api==2.5.1" --no-deps --extra-index-url https://pip.repos.neuron.amazonaws.com \
+ && ${PIP} install neuron-cc --extra-index-url https://pip.repos.neuron.amazonaws.com \
+ && ${PIP} install "tensorflow-neuron==2.5.1.*" --extra-index-url https://pip.repos.neuron.amazonaws.com
+COPY ./sagemaker /sagemaker
```
Comment (saimidu) [inline tensorflow/inference/docker/2.5/py3/sdk1.16.0/Dockerfile.neuron:108]: nit: Please add a space between the previous `RUN` statement and the `COPY` statement for readability.

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py [hunk 4]
```diff
@@ -97,11 +98,21 @@ def run_ec2_tensorflow_inference(image_uri, ec2_connection, ec2_instance_ami, gr
...
     if is_neuron:
+        #For 2.5 using rest api port instead of grpc since using curl for prediction instead of grpc
+        if str(framework_version).startswith(TENSORFLOW2_VERSION):
+            model_name= "simple"
+            model_path = os.path.join(serving_folder_path, "models", model_name)
+            src_port = "8501"
+            dst_port = "8501"
+        else:
+            src_port = grpc_port
+            dst_port = "8500"
```
Comment (saimidu) [inline test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:109]: Is the indentation correct here? Should the `else` condition apply to the outer if-condition?

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py [hunk 5]
```diff
@@ -180,6 +194,15 @@ def host_setup_for_tensorflow_inference(serving_folder_path, framework_version,
         local_scripts_path = os.path.join("container_tests", "bin", "tensorflow_serving")
         ec2_connection.run(f"mkdir -p {serving_folder_path}")
         ec2_connection.run(f"cp -r {local_scripts_path} {serving_folder_path}")
+        if is_neuron:
+            neuron_local_model = os.path.join("$HOME", "container_tests/bin/neuron_tests/simple")
```
Comment (saimidu) [inline test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:198]: `os.path.join` can take any number of input args. So please change this to:
```suggestion
            neuron_local_model = os.path.join("$HOME", "container_tests", "bin", "neuron_tests", "simple")
```

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py [hunk 6]
```diff
@@ -180,6 +194,15 @@ def host_setup_for_tensorflow_inference(serving_folder_path, framework_version,
...
+            neuron_model_dir = os.path.join(serving_folder_path, f"models/")
```
Comment (saimidu) [inline test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:199]: Please remove the `/` at the end of `f"models/"`. Also, `f"models/"` doesn't need to be an f-string, because it doesn't replace any variables in the string. It can be changed to `"models"`.

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py [hunk 7]
```diff
@@ -180,6 +194,15 @@ def host_setup_for_tensorflow_inference(serving_folder_path, framework_version,
...
+            neuron_model_file_path = os.path.join(serving_folder_path, f"models/{model_name}/1")
```
Comment (saimidu) [inline test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:200]: Please change to:
```suggestion
            neuron_model_file_path = os.path.join(serving_folder_path, "models", model_name, "1")
```

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py [hunk 8]
```diff
@@ -180,6 +194,15 @@ def host_setup_for_tensorflow_inference(serving_folder_path, framework_version,
...
+            neuron_model_file = os.path.join(neuron_model_file_path, "saved_model.pb")
```
Comment (saimidu) [inline test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:201]: `neuron_model_file` isn't being used here. Is this intentional?

File: test/dlc_tests/sanity/test_safety_check.py [hunk 9]
```diff
@@ -203,15 +203,13 @@ def _get_latest_package_version(package):
...
-        "Test will be executed in the 'mainline' pipeline and canaries pipeline."
-    ),
-)
+# @pytest.mark.skipif(
+#     not (is_mainline_context() or (is_canary_context() and is_time_for_canary_safety_scan())),
+#     reason=(
+#         "Skipping the test to decrease the number of calls to the Safety Check DB. "
+#         "Test will be executed in the 'mainline' pipeline and canaries pipeline."
+#     )
+# )
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/dlc_tests/sanity/test_safety_check.py:214]: @aws-vrnatham Please undo this comment, and continue to use `safety_check_test = true` in the config toml file to allow the safety test to run.

File: test/test_utils/__init__.py [hunk 10]
```diff
@@ -467,7 +468,7 @@ def request_pytorch_inference_densenet(
 
 
 @retry(stop_max_attempt_number=20, wait_fixed=10000, retry_on_result=retry_if_result_is_false)
-def request_tensorflow_inference(model_name, ip_address="127.0.0.1", port="8501", inference_string = "'{\"instances\": [1.0, 2.0, 5.0]}'"):
+def request_tensorflow_inference(model_name, ip_address="127.0.0.1", port="8501", connection = None, inference_string="'{\"instances\": [1.0, 2.0, 5.0]}'"):
```
Comment (saimidu) [inline test/test_utils/__init__.py:471]: Please change this to:
```suggestion
def request_tensorflow_inference(
    model_name,
    ip_address="127.0.0.1",
    port="8501",
    inference_string="'{\"instances\": [1.0, 2.0, 5.0]}'",
    connection=None,
):
```
so that the positional order of the inputs is still preserved after this change (in case any other place where this function is used depends on positional args), and so that this line complies with formatting requirements.

File: test/dlc_tests/sanity/test_safety_check.py [hunk 11]
```diff
@@ -201,13 +201,11 @@ def _get_latest_package_version(package):
...
 @pytest.mark.skipif(
-    not (
-        is_mainline_context() or is_safety_test_context() or (is_canary_context() and is_time_for_canary_safety_scan())
-    ),
+    not (is_mainline_context() or (is_canary_context() and is_time_for_canary_safety_scan())),
     reason=(
         "Skipping the test to decrease the number of calls to the Safety Check DB. "
         "Test will be executed in the 'mainline' pipeline and canaries pipeline."
-    ),
+    )
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_safety_check.py:208]: Please revert these changes completely.

File: tensorflow/buildspec-neuron.yml (hunk 1 above)
Comment (aws-vrnatham) [reply tensorflow/buildspec-neuron.yml:37]: The actual size is 8.6GB and hence had it as 10GB.

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py (hunk 4 above)
Comment (aws-vrnatham) [reply test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:109]: Yes. it is correct.. #For 2.5 using rest api port instead of grpc since using curl for prediction instead of grpc.

File: test/test_utils/__init__.py [hunk 12]
```diff
@@ -1065,7 +1071,7 @@ def get_framework_and_version_from_tag(image_uri):
             "2.2.3": "2.2.3.2.0.3.0",
             "2.3.4": "2.3.4.2.0.3.0",
             "2.4.3": "2.4.3.2.0.3.0",
-            "2.5.1": "2.5.1.2.0.3.0",
+            "2.5.1": "2.5.0.2.0.3.0",
```
Comment (saimidu) [inline test/test_utils/__init__.py]: Why was this change reverted?

Comment (aws-vrnatham) [reply test/test_utils/__init__.py]: oh ok..this was a mistake while doing the merge from the latest master. Will fix it. Thanks

**Review by saimidu — DISMISSED** [review]
Approved. Please revert config toml changes.

**Review by saimidu — APPROVED** [review]
